fendermint/vm/topdown/src/voting.rs

```diff
@@ -0,0 +1,249 @@
+            }
+        }
+
+        chain.insert(block_height, block_hash);
+
+        self.chain.write(chain)?;
+
+        Ok(())
+    }
+
+    /// Add a vote we received.
+    ///
+    /// Returns `true` if this vote was added, `false` if it was ignored as a
+    /// duplicate or a height we already finalized, and an error if it's an
+    /// equivocation or from a validator we don't know.
+    pub fn add_vote(
+        &self,
+        validator_key: K,
+        block_height: BlockHeight,
+        block_hash: V,
+    ) -> StmResult<bool, Error<K, V>> {
+        let min_height = self.last_finalized_height()?;
+
+        if block_height < min_height {
+            return Ok(false);
+        }
+
+        if !self.known_validator(&validator_key)? {
+            return abort(Error::UnknownValidator(validator_key));
+        }
+
+        let mut votes = self.votes.read_clone()?;
+        let votes_at_height = votes.entry(block_height).or_default();
+
+        for (bh, vs) in votes_at_height.iter() {
+            if *bh != block_hash && vs.contains(&validator_key) {
+                return abort(Error::Equivocation(
+                    validator_key,
+                    block_height,
+                    block_hash,
+                    bh.clone(),
+                ));
+            }
+        }
+
+        let votes_for_block = votes_at_height.entry(block_hash).or_default();
+
+        if votes_for_block.insert(validator_key).is_some() {
+            return Ok(false);
+        }
+
+        if *self.pause_votes.read()? {
+            retry()?;
+        }
+
+        self.votes.write(votes)?;
+
+        Ok(true)
+    }
+
+    /// Pause adding more votes until we are finished calling `find_quorum` which
+    /// automatically re-enables them.
+    pub fn pause_votes_until_find_quorum(&self) -> Stm<()> {
+        self.pause_votes.write(true)
+    }
+
+    /// Find a block on the (from our perspective) finalized chain that gathered enough votes from validators.
+    pub fn find_quorum(&self) -> Stm<Option<(BlockHeight, V)>> {
+        self.pause_votes.write(false)?;
+
+        let quorum_threshold = self.quorum_threshold()?;
+        let chain = self.chain.read()?;
+
+        let Some((finalized_height, _)) = chain.get_min() else {
+            return Ok(None);
+        };
+
+        let votes = self.votes.read()?;
+        let power_table = self.power_table.read()?;
+
+        let mut weight = 0;
+        let mut voters = im::HashSet::new();
+
+        for (block_height, block_hash) in chain.iter().rev() {
+            if block_height == finalized_height {
+                break;
+            }
+            // Skipping null blocks
+            if let Some(block_hash) = block_hash {
+                if let Some(votes_at_height) = votes.get(block_height) {
+                    if let Some(votes_for_block) = votes_at_height.get(block_hash) {
+                        for vk in votes_for_block {
```

Comment:
nit: check for None instead to keep code less indented for readability

Reply:
Changed it to `let Some(...) else { continue; }` to reduce indentation, let me know if this works for you.